extensions/itunes/src/album.tsx

```diff
@@ -0,0 +1,253 @@
+        <ActionPanel>
+          <PushAction
+            title={`Play the Album`}
+            target={
+              <RunScript
+                album={searchResult.album}
+                artist={searchResult.artist}
+              />
+            }
+            icon={Icon.ArrowRight}
+          />
+        </ActionPanel>
+      }
+    />
+  );
+}
+
+function useSearch() {
+  const [state, setState] = useState<SearchState>({
+    results: [],
+    isLoading: true,
+  });
+  const cancelRef = useRef<AbortController | null>(null);
+
+  useEffect(() => {
+    search("");
+    return () => {
+      cancelRef.current?.abort();
+    };
+  }, []);
+
+  async function search(searchText: string) {
+    cancelRef.current?.abort();
+    cancelRef.current = new AbortController();
+    try {
+      setState((oldState) => ({
+        ...oldState,
+        isLoading: true,
+      }));
+      await performSearch(searchText, cancelRef.current.signal);
+      setState((oldState) => ({
+        ...oldState,
+        results: match.map((entry) => {
+          return {
+            id: randomId(),
+            ...entry,
+          };
+        }),
+        isLoading: false,
+      }));
+    } catch (error) {
+      if (error instanceof AbortError) {
+        return;
+      }
+      console.error("search error", error);
+      showToast(
+        ToastStyle.Failure,
+        "Could not perform search",
+        String(error)
+      ).then();
+    }
+  }
+
+  return {
+    state: state,
+    search: search,
+  };
+}
+
+async function performSearch(searchText: string, signal: AbortSignal) {
+  if (searchText.length < 2) return;
+
+  let sherlock = false;
+  let dictBlock = false;
+  let buffer = {
+    album: "",
+    artist: "",
+  };
+  match = [];
+  const read = readline.createInterface({
+    input: fs.createReadStream(XMLPath),
+    output: process.stdout,
+    terminal: false,
+  });
+  read.on("line", (line) => {
+    if (line.indexOf("</dict>") >= 0) {
+      dictBlock = false;
+      if (
+        match.filter((entry) => {
+          return entry.artist === buffer.artist && entry.album === buffer.album;
+        }).length === 0 &&
+        buffer.album.length > 0 &&
+        buffer.album.toLowerCase().indexOf(searchText.toLowerCase()) >= 0
+      ) {
+        match = [...match, buffer];
+      }
+      buffer = {
+        album: "",
+        artist: "",
+      };
+    }
+    if (line.indexOf("<dict>") >= 0) {
+      dictBlock = true;
+      return;
+    }
+    if (dictBlock) {
+      const [key, value] = [
+        line.match(/(?<=<key>).*?(?=<\/key>)/) || [],
+        line.match(/(?<=<string>).*?(?=<\/string>)/) || "",
+      ];
+      if (key[0] === "Artist") {
+        buffer.artist = value[0];
+      }
+      if (key[0] === "Album") {
+        buffer.album = value[0];
+      }
+    }
+  });
+  read.on("close", () => {
+    match = Array.from(new Set(match));
+    sherlock = true;
+  });
+  await waitUntil(() => sherlock || signal.aborted);
```

Comment:
IMO this can be rewritten as a single promise:
```js
return new Promise((resolve, reject) => {
  const read = readline.createInterface({...})
  ...
  read.on('close', () => {
    resolve(Array.from(new Set(match)))
  })

  // probably need to handle read.on('error') as well

  signal.addEventListener('abort', () => {
    reject(new AbortError('cancelled'))
  })
})
```

that way your `performSearch` returns a promise which resolves the search results and you don't have to poll or use a global variable